src/runtime/methodbinder.cs

```diff
@@ -275,19 +282,40 @@ internal static MethodInfo MatchSignatureAndParameters(MethodInfo[] mi, Type[] g
 
                     for (int n = 0; n < clrnargs; n++) {
                         IntPtr op;
-                        if (n < pynargs)
-                        {
-                            if (arrayStart == n)
-                            {
+                        if (n < pynargs) {
+                            if (arrayStart == n) {
                                 // map remaining Python arguments to a tuple since
                                 // the managed function accepts it - hopefully :]
                                 op = Runtime.PyTuple_GetSlice(args, arrayStart, pynargs);
                             }
-                            else
-                            {
+                            else {
                                 op = Runtime.PyTuple_GetItem(args, n);
                             }
-                            Type type = pi[n].ParameterType;
+                            type = null;
+                            if  (_methods.Length>1) { 
+                                IntPtr pyoptype = IntPtr.Zero;
+                                try {
+                                    pyoptype = Runtime.PyObject_Type(op);
+                                }
+                                catch (System.Exception e) {
```

Comment:
@tonyroberts yes, PyObject_Type can throw SystemError,  this is from Python docs:

PyObject* PyObject_Type(PyObject *o)
Return value: New reference.
When o is non-NULL, returns a type object corresponding to the object type of object o. On failure, raises SystemError and returns NULL. This is equivalent to the Python expression type(o). This function increments the reference count of the return value. There’s really no reason to use this function instead of the common expression o->ob_type, which returns a pointer of type PyTypeObject*, except when the incremented reference count is needed.